src/project.coffee

```diff
@@ -39,6 +39,20 @@ class Project extends Model
     @emitter = new Emitter
     @buffers ?= []
 
+    # Mapping from the real path of a {Directory} to a {Promise} that resolves
+    # to either a {Repository} or null. Ideally, the {Directory} would be used
+    # as the key; however, there can be multiple {Directory} objects created for
+    # the same real path, so it is not a good key.
```

Comment:
I'm not sure if `Project` should be maintaining that map or if a `DirectoryProvider` should be responsible for its own cache.

Renamed to `repositoryPromisesByDirectory`, as requested. One thing to note is that I made this a `Map` rather than a `{}` so that there would be fewer code changes if the key type changed (hence having `Map` in the name). However, I'm not sure if we'll every be confident that there will be exactly one `Directory` for path, in which case the key will probably always be a string and I can just change this to a `{}` now.